### PR TITLE
perf: reduce function calls for shallower stacks

### PR DESCRIPTION
**Pull request by jquense**
starts to address #1017 bringing the benchmark for sync and async within striking distance of joi at least

### CODE REVIEW

File: src/util/createValidation.js [hunk 1]
```diff
@@ -1,94 +1,41 @@
...
-      new ValidationError(formatError(message, params), value, path, type),
-      { params },
-    );
-  };
-}
-
-export default function createValidation(options) {
-  let { name, message, test, params } = options;
 
+export default function createValidation(config) {
```
Comment (jquense) [inline src/util/createValidation.js:5]: most wins due to simplifying this call by reducing indirection and inlining helpers

File: src/util/runTests.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
+import ValidationError from '../ValidationError';
+import { once } from './async';
+
+export default function runTests(options, cb) {
```
Comment (jquense) [inline src/util/runTests.js:4]: more big wins here, by making this function less clear with the loop but much shallower

**Comment by JaneJeon** [comment]
This is excellent! I'm getting 96k req/s with the synchronous implementation (vs. joi's 157k) - well within striking range of joi (50%), _and_ fast enough to not be the bottleneck for either express or fastify!

I'd love to see further work on this, and I'm willing to even contribute performance regression tests if necessary. Thanks for doing this!

**Comment by jquense** [comment]
> and I'm willing to even contribute performance regression tests if necessary

That would be excellent and appreciated, having some benchmarks in the repo would be really helpful for tracking this stuff longer term. Doesn't need to be fancy just some numbers for yup to compare new versions against

**Comment by JaneJeon** [comment]
will get on it after work!

**Pull request merged**